Add default hooks to save tensors on CPU

Varal7 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62410
* __->__ #61928

Fix #57100.
Creates a context-manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are moved* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the user so desires, we move the tensor to `pin_memory` during packing so that the unpacking can be done asynchronously.

With the current PR, hooks are registered globally, across threads. In the near future, we want to make these hooks thread-local.

See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279) and [note about training large models](https://github.com/pytorch/pytorch/pull/61928#issuecomment-887009448)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)

Varal7 committed:
Add context manager to save tensors on CPU

Fix #57100

[ghstack-poisoned]

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a context manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are actually copied to cpu, then copied back to the appropriate device for the backward pass.


[ghstack-poisoned]

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a context manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are actually copied to cpu, then copied back to the appropriate device for the backward pass.


[ghstack-poisoned]

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a context manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are actually copied to cpu, then copied back to the appropriate device for the backward pass.


[ghstack-poisoned]

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a context manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are actually copied to cpu, then copied back to the appropriate device for the backward pass.


[ghstack-poisoned]

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a context manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are actually copied to cpu, then copied back to the appropriate device for the backward pass.


[ghstack-poisoned]

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a context manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are actually copied to cpu, then copied back to the appropriate device for the backward pass.


[ghstack-poisoned]

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a context manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are actually copied to cpu, then copied back to the appropriate device for the backward pass.


[ghstack-poisoned]

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a context manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are actually copied to cpu, then copied back to the appropriate device for the backward pass.


[ghstack-poisoned]

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a context manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are actually copied to cpu, then copied back to the appropriate device for the backward pass.


[ghstack-poisoned]

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a context manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are actually copied to cpu, then copied back to the appropriate device for the backward pass.


[ghstack-poisoned]

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a context manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are actually copied to cpu, then copied back to the appropriate device for the backward pass.


[ghstack-poisoned]

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a context manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are actually copied to cpu, then copied back to the appropriate device for the backward pass.


[ghstack-poisoned]

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a context manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are actually copied to cpu, then copied back to the appropriate device for the backward pass.


[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848526).

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a context manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are actually copied to cpu, then copied back to the appropriate device for the backward pass.

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)

[ghstack-poisoned]

Varal7 commented:
Benchmark:
```python
def repeat(f, x, n):
    for _ in range(n):
        x = f(x)
    return x

def get_stats(n, save_on_cpu):
    a = torch.ones(1, device="cuda", requires_grad=True)
    start = time.time()
    torch.autograd.graph.set_save_on_cpu(save_on_cpu)
    y = repeat(lambda x: x * x, a, n)
    fw = time.time()
    torch.autograd.graph.set_save_on_cpu(False)
    y.sum().backward()
    bw = time.time()
    peak = (torch.cuda.memory_stats()['allocated_bytes.all.peak'])
    torch.cuda.reset_peak_memory_stats()
    return (n, save_on_cpu, fw - start, bw - fw, peak)
```

Full table: https://docs.google.com/spreadsheets/d/1dR4cLzj4oGA2U-xaKVY2GyJe0AhykH3imCMGLduRrNo/edit?usp=sharing 

Summary:

![image](https://user-images.githubusercontent.com/8019486/126682173-e30b9463-1853-405f-870c-ff3ff11843d8.png)

<img width="459" alt="Screen Shot 2021-07-22 at 1 24 03 PM" src="https://user-images.githubusercontent.com/8019486/126682112-c8b2bc60-140f-4d3a-b3cd-92947b92faad.png">

| n      | fw slowdown        | bw slowdown        | mem usage             |
|--------|--------------------|--------------------|-----------------------|
| 1      | 5.076963626778702  | 0.9573850868233074 | 1.1428571428571428    |
| 10     | 4.834049739025118  | 1.952815829528211  | 0.6                   |
| 100    | 5.120109257345501  | 2.155610999620339  | 0.08571428571428572   |
| 1000   | 2.8076930992611526 | 1.669780703276048  | 0.008955223880597015  |
| 10000  | 4.132995644710133  | 2.1410211547704403 | 0.0008995502248875562 |
| 100000 | 4.761377533533761  | 2.24638000636119   | 8.999550022498874e-05 |

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a context manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are actually copied* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we copy the tensor to `pin_memory` so that the unpack can be done asynchronously.


See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)

[ghstack-poisoned]

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a context manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are actually copied* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we copy the tensor to `pin_memory` so that the unpack can be done asynchronously.


See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848526).

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a context manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are actually copied* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we copy the tensor to `pin_memory` so that the unpack can be done asynchronously.


See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848526).

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a context manager `torch.autograd.graph.save_on_cpu()` under which all tensors saved during the forward pass are actually copied* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we copy the tensor to `pin_memory` so that the unpack can be done asynchronously.


See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848526).

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a function `torch.autograd.graph.set_save_on_cpu()` which can be used to register default hooks under which all tensors saved during the forward pass are actually copied* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we copy the tensor to `pin_memory` during packing so that the unpacking can be done asynchronously.


See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848526).

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a function `torch.autograd.graph.set_save_on_cpu()` which can be used to register default hooks under which all tensors saved during the forward pass are actually copied* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we copy the tensor to `pin_memory` during packing so that the unpacking can be done asynchronously.


See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848526).

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a function `torch.autograd.graph.set_save_on_cpu()` which can be used to register default hooks under which all tensors saved during the forward pass are actually copied* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we copy the tensor to `pin_memory` during packing so that the unpacking can be done asynchronously.


See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848526).

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a function `torch.autograd.graph.set_save_on_cpu()` which can be used to register default hooks under which all tensors saved during the forward pass are actually copied* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we copy the tensor to `pin_memory` during packing so that the unpacking can be done asynchronously.


See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848526).

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a function `torch.autograd.graph.set_save_on_cpu()` which can be used to register default hooks under which all tensors saved during the forward pass are actually copied* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we copy the tensor to `pin_memory` during packing so that the unpacking can be done asynchronously.


See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)

[ghstack-poisoned]

Varal7 committed:
Update on "Add context manager to save tensors on CPU"


Fix #57100.
Creates a function `torch.autograd.graph.set_save_on_cpu()` which can be used to register default hooks under which all tensors saved during the forward pass are actually copied* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we copy the tensor to `pin_memory` during packing so that the unpacking can be done asynchronously.


See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)

[ghstack-poisoned]

zou3519 commented on torch/autograd/saved_variable_default_hooks.py:
```diff
@@ -5,3 +5,19 @@ def set_saved_tensors_default_hooks(pack_hook, unpack_hook):
+            tensor.size(),
+            dtype=tensor.dtype,
+            layout=tensor.layout,
+            pin_memory=torch.cuda.is_available())
```
What is the rationale for using pin_memory? That the transfers will be faster?

The problem with pinned memory is that there is generally a limited amount of it and for larger models we'll probably run out of pinned memory and raise an error here. A good sanity check might be something like running resnet18 training

Varal7 replied:
Yes. It's so that we can use a non-blocking copy from CPU to GPU. See https://github.com/pytorch/pytorch/pull/61928#issuecomment-887009448 for an example on resnet152.

One option I like would be to take as argument a `pin_memory` bool so the user can decide whether they want pinned memory or not.

zou3519 replied:
Conventional wisdom tells me that `pin_memory` should not be used by default because it is a limited resource. I agree a pin_memory book would be nice for the user to decide if they want pinned memory or not

zou3519 commented on torch/autograd/saved_variable_default_hooks.py:
```diff
@@ -5,3 +5,19 @@ def set_saved_tensors_default_hooks(pack_hook, unpack_hook):
 def reset_saved_tensors_default_hooks():
     torch._C._autograd._reset_default_hooks()
+
+def set_save_on_cpu_hooks():
```
The PR body says this is supposed to be a context manager but it doesn't look like one

Varal7 replied:
Good catch, there should not be mentions of context managers anymore.

Varal7 committed:
Update on "Add default hooks to save tensors on CPU"


Fix #57100.
Creates a function `torch.autograd.graph.set_save_on_cpu()` which can be used to register default hooks under which all tensors saved during the forward pass are actually copied* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we copy the tensor to `pin_memory` during packing so that the unpacking can be done asynchronously.


See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848526).

Varal7 committed:
Update on "Add default hooks to save tensors on CPU"


Fix #57100.
Creates a function `torch.autograd.graph.set_save_on_cpu()` which can be used to register default hooks under which all tensors saved during the forward pass are actually copied* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we copy the tensor to `pin_memory` during packing so that the unpacking can be done asynchronously.


See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848526).

Varal7 commented:
About training real-life large models.

I tried training a resnet152 on [Imagenette](https://github.com/fastai/imagenette) on an RTX A6000.

Typical training using a **batch size of 256** uses up to **48GB** of GPU memory and is unfeasible on smaller GPUs.

Using the `save_on_cpu` hook, the GPU memory usage falls all the way down to **5760MB** for a batch size of 256.
Using pinned memory allows the `unpacking` operation to be performed in a non-blocking way so we can achieve around **14 seconds** for 10 batches during training (so 2560 images).
In contrast, not using pinned memory requires the unpacking to be asynchronous, which increases the time to process 10 training batches to around **60 seconds**.

As a fair comparison, I also tried training with a batch size of 16 (and no hooks) for which the GPU memory requirement is below 5GB. The smaller batch size is much faster with only **4 seconds** to process the same number (2560) of images.

Varal7 committed:
Update on "Add default hooks to save tensors on CPU"


Fix #57100.
Creates a function `torch.autograd.graph.set_save_on_cpu_hooks()` which can be used to register default hooks under which all tensors saved during the forward pass are actually copied* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we copy the tensor to `pin_memory` during packing so that the unpacking can be done asynchronously.


See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279) and [note about training large models](https://github.com/pytorch/pytorch/pull/61928#issuecomment-887009448)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848526).

Varal7 committed:
Update on "Add default hooks to save tensors on CPU"


Fix #57100.
Creates a function `torch.autograd.graph.set_save_on_cpu_hooks()` which can be used to register default hooks under which all tensors saved during the forward pass are actually copied* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we copy the tensor to `pin_memory` during packing so that the unpacking can be done asynchronously.


See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279) and [note about training large models](https://github.com/pytorch/pytorch/pull/61928#issuecomment-887009448)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)

[ghstack-poisoned]

Varal7 committed:
Update on "Add default hooks to save tensors on CPU"


Fix #57100.
Creates a function `torch.autograd.graph.set_save_on_cpu_hooks()` which can be used to register default hooks under which all tensors saved during the forward pass are actually copied* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we copy the tensor to `pin_memory` during packing so that the unpacking can be done asynchronously.


See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279) and [note about training large models](https://github.com/pytorch/pytorch/pull/61928#issuecomment-887009448)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)

[ghstack-poisoned]

zou3519 commented on torch/autograd/saved_variable_default_hooks.py:
```diff
@@ -5,3 +5,22 @@ def set_saved_tensors_default_hooks(pack_hook, unpack_hook):
 def reset_saved_tensors_default_hooks():
     torch._C._autograd._reset_default_hooks()
+
+def set_save_on_cpu_hooks(pin_memory=False):
```
A few questions:
- Should the design of this be a context manager instead of a global flag?
- I remember there are talks about changing hooks to be thread-local: when they are thread-local, does it make more sense for this to be a context manager?
- Is `set_save_on_cpu_hooks` the final API we are exposing to users?

zou3519 replied:
A reason why we might want this to be a context manager is, maybe the user only wants to move part of their model's saved variables to CPU

Varal7 replied:
Yes ideally, we'd want this to be a a context manager and be thread-local. For now, exposing a context manager could trick the users into thinking that the code is thread-local while it is not. The future API would be:
```python
with torch.autograd.graph.save_on_cpu():
    pass
````

For now, yes, the current API exposed to users would be `set_save_on_cpu_hooks` to set the hooks and `reset_saved_tensors_default_hooks` to clear them. (Working on docstrings right now! could incorporate in this PR if needed)

zou3519 replied:
No need to incorporate in this PR, I was curious about the plan.

How soon were you thinking we would change this to be a context manager? If we release this feature as-is in PyTorch 1.10, and then we change this feature to be a context manager in PyTorch 1.11, then we would actually need to do a deprecation cycle and that would not be ideal.

If on the other hand, the plan is to change this to be a context manager before PyTorch 1.10, then everything works out. Maybe this is more of a question for Alban?

Varal7 replied:
I'm hoping for an "everything-works-out" scenario but I need to check with Alban that the timelines align.
My PSC is next week and I'm not aiming to have landed the thread-local feature by then. However it would be my first task once I'm done with my PSC tasks. So I expect to start working on it mid-next week?

Varal7 replied:
Also, I'm not sure we actually want to depreciate the `set_save_on_cpu_hooks` API.
Similarly to grad mode that has `torch.set_grad_enabled(False)` and the context-manager `with torch.no_grad(): pass`

zou3519 replied:
Offering something to modify global state seems like a good way for the user to shoot themselves in the foot. Let's say:
- the user imports a model that they copy-pasted online that sets cpu hooks
- they don't end up running the model and instead run a different model (that didn't globally set cpu hooks)
- Then the user may be confused to find that their new model saves activations to CPU!
- The user might not even notice this -- they might notice a performance drop and be confused.

`torch.set_grad_enabled` is a context manager (https://pytorch.org/docs/stable/generated/torch.set_grad_enabled.html), not global state, so it avoids this problem.

My opinion is that `set_save_on_cpu_hooks` (and other similar APIs we offer directly to the user) should always be context managers, even if they're not thread-local.

albanD replied:
> torch.set_grad_enabled is a context manager (https://pytorch.org/docs/stable/generated/torch.set_grad_enabled.html), not global state, so it avoids this problem.

Well it is both haha You can use it as a context manager or as a global setter
https://github.com/pytorch/pytorch/blob/093495d3f0985fbe2aea58044798d588c0fe6889/torch/autograd/grad_mode.py#L179

I don't see a reason not to have a functional API for this. It is convenient for debugging purposed in particular or for users that want to change the default across their code.
But I do agree that it is not the best composability-wise and we should not encourage it as the main way to do things.

Varal7 replied:
I'm weighing the pros and cons of the two options.
Your example makes a strong point in favor of using a context-manager even when the operation is not thread local.

The con of using a context-manager in that case is it might cause surprising errors to the user.

For instance, this fails (while we are not thread-local):

```python
class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.a = nn.Parameter(torch.randn(10))

    def forward(self, x):
        with torch.autograd.graph.save_on_cpu():
            y = self.a * (self.a * x)
        return y

net = torch.nn.DataParallel(Model())
x = torch.ones(10, 10)
y = net(x)
```

But this works:

```python
class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.a = nn.Parameter(torch.randn(10))

    def forward(self, x):
        y = self.a * (self.a * x)
        return y

net = torch.nn.DataParallel(Model())
x = torch.ones(10, 10)
with torch.autograd.graph.save_on_cpu():
    y = net(x)
```

If we were to only keep one API, I'm in favor of the context manager because at least the error will explicitly say "Setting default hooks but they are already defined", while your example shows that leaving an "unclosed" `torch.autograd.graph.set_save_on_cpu_hooks()` can cause sneaky errors.

===

A note regarding `torch.set_grad_enabled`: it's both a context-manager and a function you can call "in the void".
And by the way, I'm not fan of the way it was implemented to allow both usages: the grad mode is set on `__init__` instead of on `__enter__`.

So for instance:

```python
x = torch.ones(5, requires_grad=True)
no_grad_ctx = torch.set_grad_enabled(False)
a = x * x
with no_grad_ctx:
    b = x * x
print(a.requires_grad) # False
print(b.requires_grad) # False
```

zou3519 replied:
My bad, I didn't realize `torch.set_grad_enabled` could also be used as a function!
```
class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.a = nn.Parameter(torch.randn(10))

    def forward(self, x):
        with torch.autograd.graph.save_on_cpu():
            y = self.a * (self.a * x)
        return y

net = torch.nn.DataParallel(Model())
x = torch.ones(10, 10)
y = net(x)
```
What happens with this today?

Varal7 replied:
On a machine with only one device, it would work as expected.

On a machine with multiple devices (e.g. more than one GPU), it would raise
`RuntimeError: Setting default hooks but they have already been set. Hint: only one pair of hooks is allowed at a time.`

Varal7 replied:
I made the change to context-manager.

zou3519 reviewed (commented):
The benchmarks look really good! I had some questions about the design of this and what the plan for things are

Varal7 committed:
Update on "Add default hooks to save tensors on CPU"


Fix #57100.
Creates a function `torch.autograd.graph.set_save_on_cpu_hooks()` which can be used to register default hooks under which all tensors saved during the forward pass are actually copied* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we copy the tensor to `pin_memory` during packing so that the unpacking can be done asynchronously.

With the current PR, hooks are set with `torch.autograd.graph.set_save_on_cpu_hooks()` and unset with `torch.autograd.graph.reset_saved_tensors_default_hooks`. In the near future, we want to make these hooks thread-local and expose a context manager `torch.autograd.graph.save_on_cpu`.

See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279) and [note about training large models](https://github.com/pytorch/pytorch/pull/61928#issuecomment-887009448)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)



[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848526).

Varal7 committed:
Update on "Add default hooks to save tensors on CPU"


Fix #57100.
Creates a function `torch.autograd.graph.set_save_on_cpu_hooks()` which can be used to register default hooks under which all tensors saved during the forward pass are moved* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we move the tensor to `pin_memory` during packing so that the unpacking can be done asynchronously.

With the current PR, hooks are set with `torch.autograd.graph.set_save_on_cpu_hooks()` and unset with `torch.autograd.graph.reset_saved_tensors_default_hooks`. In the near future, we want to make these hooks thread-local and expose a context manager `torch.autograd.graph.save_on_cpu`.

See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279) and [note about training large models](https://github.com/pytorch/pytorch/pull/61928#issuecomment-887009448)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)



[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848526).

Varal7 committed:
Update on "Add default hooks to save tensors on CPU"


Fix #57100.
Creates a function `torch.autograd.graph.set_save_on_cpu_hooks()` which can be used to register default hooks under which all tensors saved during the forward pass are moved* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we move the tensor to `pin_memory` during packing so that the unpacking can be done asynchronously.

With the current PR, hooks are set with `torch.autograd.graph.set_save_on_cpu_hooks()` and unset with `torch.autograd.graph.reset_saved_tensors_default_hooks`. In the near future, we want to make these hooks thread-local and expose a context manager `torch.autograd.graph.save_on_cpu`.

See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279) and [note about training large models](https://github.com/pytorch/pytorch/pull/61928#issuecomment-887009448)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)



[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848526).

Varal7 committed:
Update on "Add default hooks to save tensors on CPU"


Fix #57100.
Creates a function `torch.autograd.graph.set_save_on_cpu_hooks()` which can be used to register default hooks under which all tensors saved during the forward pass are moved* to cpu, then copied back to the appropriate device for the backward pass.

*If the tensor was already on cpu, the entire operation is a no op.

If the tensor is on GPU, we move the tensor to `pin_memory` during packing so that the unpacking can be done asynchronously.

With the current PR, hooks are set with `torch.autograd.graph.set_save_on_cpu_hooks()` and unset with `torch.autograd.graph.reset_saved_tensors_default_hooks`. In the near future, we want to make these hooks thread-local and expose a context manager `torch.autograd.graph.save_on_cpu`.

See [benchmark](https://github.com/pytorch/pytorch/pull/61928#issuecomment-885089279) and [note about training large models](https://github.com/pytorch/pytorch/pull/61928#issuecomment-887009448)

Differential Revision: [D29848526](https://our.internmc.facebook.com/intern/diff/D29848526)



[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29848526).

zou3519 commented on torch/autograd/saved_variable_default_hooks.py:
```diff
@@ -1,7 +1,35 @@
+            if not pin_memory:
+                return (tensor.device, tensor.cpu())
+
+            storage = torch.empty(
+                tensor.size(),
+                dtype=tensor.dtype,
+                layout=tensor.layout,
+                pin_memory=(torch.cuda.is_available() and not tensor.is_sparse))
+            storage.copy_(tensor)
```
Why can't we do `storage = tensor.to(device='cpu', pin_memory=True)`?

Also nit: we should rename storage into something else; storage can be confused with `torch.Storage`

Varal7 replied:
I don't think that `pin_memory` is an acceptable argument of `torch.tensor.to`

Ok for `storage` rename.

zou3519 replied:
> I don't think that `pin_memory` is an acceptable argument of `torch.tensor.to`

Good point, thanks for the clarification

rohan-varma replied:
would something like `storage = tensor.to("cpu", non_blocking=True).pin_memory()` work?

albanD replied:
That would mean 2 copies
